### PR TITLE
[REVIEW] Write global _metadata file for Dask-based parquet output

### LINKED ISSUE #59: [FEA] Write shared "_metadata" file for output parquet data in DaskWorkflow
<a href="https://github.com/rjzamora"></img></a> **Issue by [rjzamora](https://github.com/rjzamora)**
_Thursday May 28, 2020 at 16:43 GMT_
_Originally opened as https://github.com/rapidsai/recsys/issues/179_

----

**Is your feature request related to a problem? Please describe.**
The `DaskWorkflow` implemntation introduced in #136 does not currently write a shared parquet `"_metadata"` file when the `to_dataset` API is used.  This is because of the limitation discussed in [cudf#284](https://github.com/rapidsai/cudf/issues/5284).  Once that issue is resolved upstream, the `DaskWorkflow` code should be modified to write the shared file (having this file is a significant performance boost if/when dask is used to read the dataset back in).

**Describe the solution you'd like**
Once the upstream changes are made, metadata file creation will look very similar to [this existing cudf method](https://github.com/rapidsai/cudf/blob/56769d4755091e3fe892fcdd020bd1e2a9a19cfe/python/dask_cudf/dask_cudf/io/parquet.py#L130).

**Describe alternatives you've considered**
The alternative is to skip `"_metadata"` creation, or to create the shared file with pyarrow after the fact (requiring a followup python process to parse all the footer metadata after it is written).

### PR DESCRIPTION
**Pull request by rjzamora**
Closes #59 
Address Dask component of #51

Now that [cudf#5411](https://github.com/rapidsai/cudf/pull/5411) is merged, we can/should write the global _metadata file after writing out a (processed/shuffled) parquet dataset.  This PR covers the Dask-related changes (note that we still need to align the single/multi-gpu IO mechanisms in the future).

### CODE REVIEW

File: tests/conftest.py [hunk 1]
```diff
@@ -47,7 +47,7 @@
 def client():
     global _CLIENT
     if _CLIENT is None:
-        _CLIENT = Client(LocalCluster(n_workers=2))
+        _CLIENT = Client(LocalCluster(n_workers=2, threads_per_worker=1))
```
Comment (rjzamora) [inline tests/conftest.py:50]: Turns out we get race conditions during worker caching when we allow them to use multiple threads.

Comment (benfred) [reply tests/conftest.py:50]: Should we put a lock around the worker cache instead to handle this? (Or make the worker cache thread local?)

Comment (rjzamora) [reply tests/conftest.py:50]: Sorry - I am probably wring to call this a "race condition."  For some reason, the worker can end up creating >1 ParquetWriter objects for the same output split.  Everything seems to "work" fine, but the number of output files ends up being wrong.

Not really sure how to add a lock here.  It seems that I can avoid the issue if I add a wait and retry block after the first call to `pq_writer_cache.get` in `get_pq_writer`.  I'll try to figure out the "correct" way to ensure synchronization

Comment (rjzamora) [reply tests/conftest.py]: Okay - I am adding a lock for now, because that seems to work fine.

File: nvtabular/io.py [hunk 2]
```diff
@@ -612,17 +614,21 @@ def clean_pw_cache():
         return
 
 
+def close_cached_pw(fs):
+    md_dict = {}
+    with get_cache() as cache:
+        for path, (pw, bio) in cache.pq_writer_cache.items():
+            fn = bio.split(fs.sep)[-1]
```
Comment (benfred) [inline nvtabular/io.py:621]: Can't bio be a bytes IO object if 'mem=True' when calling get_pq_writer? Should we use path instead?

Comment (rjzamora) [reply nvtabular/io.py:621]: I can add an assertion, but this function should never be called for the "full" shuffle. bio will always be a path here.

**Pull request merged**